### PR TITLE
Rework readme

### PR DESCRIPTION
**Pull request by Zifzaf**
Changing structure of SW I2C already approved in https://github.com/Sensirion/embedded-common/pull/67 .
Added in New README.md.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,14 @@
...
-This repository contains the Sensirion I2C hardware abstraction layer,
-the Sensirion SHDLC/UART hardware abstraction layer and commonly shared code
-for both parts. It is a dependency of the Sensirion embedded drivers.
+This repository contains the implementation of the Sensirion I2C and SHDLC
+protocol and some commonly used functions for sensor implementations.
+It is a dependency of the Sensirion embedded sensor drivers. Sensirion
+senors all use a protocol on top of the bit level communications like I2C or
+UART. There are two of those kinds protocols used by Sensirion sensors, one
+for UART communication and one for I2C communication. These two work the same
+for all sensor types and therefore need to be implemented only once. To keep
```
Comment (abrauchli) [inline README.md:8]: (please reformat when applying)

```suggestion
senors all use a protocol on top of the I2C/UART communications layer.
There are two of those kinds protocols used by Sensirion sensors, one
for UART communication, called SHDLC, and one for I2C communication.
These are shared amongst sensor types and are therefore implemented only once.
To keep
```

File: README.md [hunk 2]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
-* `shdlc` folder (implementation of the SHDLC communication and UART HAL
-  including sample code for some platforms)
-* `i2c` folder (implementation of the I2C communication)
-* `i2c/hw` folder (stub for implementation of the hardware I2C HAL,
-  including sample code for some platforms)
-* `i2c/sw` folder (stub for implementation of the software I2C HAL
-  (bit-banging GPIOs), including sample code for some platforms)
-* `tests-common-config` folder (configuration files for tests, used by driver repositories)
+
+The source code in this repository can be spilt into three parts.
```
Comment (abrauchli) [inline README.md:21]:
```suggestion
The source code in this repository is spilt into three parts: Common code, I2C protocol and SHDLC protocol.
```

File: README.md [hunk 3]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+
+### Common
+
+In the `common/` folder are three files. `senirion_config.h` is a general
```
Comment (abrauchli) [inline README.md:25]:
```suggestion
In the `common/` folder are three files. `senirion_config.h` is the general
```

File: README.md [hunk 4]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+configuration file. Depending on your hardware you will need to edit this.
+You can find more about this further down. There are also
```
Comment (abrauchli) [inline README.md:27]:
```suggestion
For more information, see the section named `sensirion_config.h` further down in this document. There are also
```

File: README.md [hunk 5]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+`sensirion_common.[ch]`. These contain common helper functions used by both
+UART and I2C sensor drivers.
+
+### I2C
+
+In the `i2c/` folder is the implementation of the protocol used by Sensirion
+sensors for I2C communication. This implementation is spilt into two parts.
+The implementation of the high level protocol in the files `sensirion_i2c.[ch]`
+and a hardware abstraction layer in `senirion_i2c_hal.[ch]`.
+In `sensirion_i2c.[ch]` all the implementation details (CRC calculation,byte order,
```
Comment (abrauchli) [inline README.md:37]:
```suggestion
In `sensirion_i2c.[ch]` all the implementation details (CRC calculation, byte order swapping,
```

File: README.md [hunk 6]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+...) of Sensirion's I2C communication are implemented. The finished frame is then
+send to your hardware using the HAL (Hardware Abstraction Layer) in the other
```
Comment (abrauchli) [inline README.md:39]:
```suggestion
sent to the attached hardware using the Hardware Abstraction Layer (HAL) in the other
```

File: README.md [hunk 7]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+two files. Since the way how to send out this frame depends on underlying
+hardware and operating system we provide different implementations for that part
```
Comment (abrauchli) [inline README.md:41]:
```suggestion
hardware and operating system, we provide different implementations for that part
```

File: README.md [hunk 8]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+of the code. You can find these in the folder `i2c/sample-implementations/` for the
+most common platforms. We even provide a implementation for software I2C using
+bit-banging on GPIOs in `i2c/sample-implementations/GPIO_bit_banging/`. If your
+platform isn't there you can use the empty structure in the `i2c/` folder to
```
Comment (abrauchli) [inline README.md:45]:
```suggestion
platform isn't currently there, you can use the provided template structure in the `i2c/` folder to
```

File: README.md [hunk 9]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+implement it yourself.
+
+### SHDLC
+
+In the `shdlc` folder is the implementation of the protocol used by Sensirion
+sensors for UART communication. These sensors use Sensirions own SHDLC protocol.
+The high level implementation of this protocol can be found in
+`sensirion_shdlc.[ch]` and the hardware abstraction layer in `sensirion_uart_hal.[ch]`.
+In `sensirion_shdlc.[ch]` all the implementation details (checksum calculation,
+byte order, frame constructon, ...) of Sensirion's SHDLC communication are
```
Comment (abrauchli) [inline README.md:55]:
```suggestion
byte order swapping, frame construction, ...) of Sensirion's SHDLC communication are
```

File: README.md [hunk 10]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+implemented. The finished frame is then send to your hardware using the HAL
+(Hardware Abstraction Layer) in the other two files. Since the way how to send
```
Comment (abrauchli) [inline README.md:57]:
```suggestion
in the other two files. Since the way how to send
```

File: README.md [hunk 11]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+
+### SHDLC
+
+In the `shdlc` folder is the implementation of the protocol used by Sensirion
+sensors for UART communication. These sensors use Sensirions own SHDLC protocol.
+The high level implementation of this protocol can be found in
+`sensirion_shdlc.[ch]` and the hardware abstraction layer in `sensirion_uart_hal.[ch]`.
+In `sensirion_shdlc.[ch]` all the implementation details (checksum calculation,
+byte order, frame constructon, ...) of Sensirion's SHDLC communication are
+implemented. The finished frame is then send to your hardware using the HAL
```
Comment (abrauchli) [inline README.md:56]:
```suggestion
implemented. The finished frame is then sent to your hardware using the HAL
```

File: README.md [hunk 12]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+In `sensirion_i2c.[ch]` all the implementation details (CRC calculation,byte order,
+...) of Sensirion's I2C communication are implemented. The finished frame is then
+send to your hardware using the HAL (Hardware Abstraction Layer) in the other
+two files. Since the way how to send out this frame depends on underlying
+hardware and operating system we provide different implementations for that part
+of the code. You can find these in the folder `i2c/sample-implementations/` for the
+most common platforms. We even provide a implementation for software I2C using
+bit-banging on GPIOs in `i2c/sample-implementations/GPIO_bit_banging/`. If your
+platform isn't there you can use the empty structure in the `i2c/` folder to
+implement it yourself.
```
Comment (abrauchli) [inline README.md:46]:
```suggestion
implement it yourself. We're very happy to review and include more architectures in the form of a pull request on GitHub.
```

File: README.md [hunk 13]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+`sensirion_shdlc.[ch]` and the hardware abstraction layer in `sensirion_uart_hal.[ch]`.
+In `sensirion_shdlc.[ch]` all the implementation details (checksum calculation,
+byte order, frame constructon, ...) of Sensirion's SHDLC communication are
+implemented. The finished frame is then send to your hardware using the HAL
+(Hardware Abstraction Layer) in the other two files. Since the way how to send
+out this frame depends on underlying hardware and operating system we provide
+different implementations for that part of the code. You can find these in the
+folder `shdlc/sample-implementaions/` for the most common platforms. If your
+platform isn't there you can use the empty structure in the `shdlc/` folder to
+implement it yourself.
```
Comment (abrauchli) [inline README.md:62]:
```suggestion
implement it yourself. We're very happy to review and include more architectures in the form of a pull request on GitHub.
```

File: README.md [hunk 14]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+The high level implementation of this protocol can be found in
+`sensirion_shdlc.[ch]` and the hardware abstraction layer in `sensirion_uart_hal.[ch]`.
+In `sensirion_shdlc.[ch]` all the implementation details (checksum calculation,
+byte order, frame constructon, ...) of Sensirion's SHDLC communication are
+implemented. The finished frame is then send to your hardware using the HAL
+(Hardware Abstraction Layer) in the other two files. Since the way how to send
+out this frame depends on underlying hardware and operating system we provide
+different implementations for that part of the code. You can find these in the
+folder `shdlc/sample-implementaions/` for the most common platforms. If your
+platform isn't there you can use the empty structure in the `shdlc/` folder to
```
Comment (abrauchli) [inline README.md:61]:
```suggestion
platform isn't currently there, you can use the provided template structure in the `shdlc/` folder to
```

File: README.md [hunk 15]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+to be made according to your hardware platform.
 
 ### sensirion\_config.h
-You need to specify the integer sizes on your system.
-We assume you have the `<stdint.h>` file available, so we have an
-include directive for it. If this is not the case on your platform you need to
-specify the following types yourself:
+
+In this file we keep all the included libraries for our drivers. It is located
+in the `common/` folder.  The two libraries our drivers use are `<stdint.h>` and
```
Comment (abrauchli) [inline README.md:73]:
```suggestion
in the `common/` folder. The two libraries our drivers use are `<stdint.h>` and
```

File: README.md [hunk 16]
```diff
@@ -10,21 +17,64 @@ files) from the respective projects. An overview is listed at
...
+`<stdlib.h>`. The features we use out of those libraries is standard integer sizes
+from `<stdint.h>` and NULL from `<stdlib.h>`. We assume you have both those libraries
```
Comment (abrauchli) [inline README.md:75]:
```suggestion
from `<stdint.h>` and `NULL` from `<stdlib.h>`. We assume that those libraries are provided by your toolchain, compiler or are
```

File: README.md [hunk 17]
```diff
@@ -35,25 +85,31 @@ specify the following types yourself:
...
-bit-banging GPIOs. You will thus need to implement the stubs necessary to drive
-the GPIO pins used for SDA (data line) and SCL (clock line) along with the delay
-function, that must be precise enough to drive the pins at the desired
-frequency.  For further information please check out some of the sample
-implementations in the `i2c/sw/sample-implementations` folder in this
-repository.
+### sensirion\_i2c\_hal.c
+
+As already explained earlier is all the I2C related code in this repository based
+on Sensirion's I2c HAL. It is defined in `sensirion_i2c_hal.h` and implemented in
```
Comment (abrauchli) [inline README.md:97]:
```suggestion
on Sensirion's I2C HAL. It is defined in `sensirion_i2c_hal.h` and implemented in
```

File: i2c/README.md [hunk 18]
```diff
@@ -0,0 +1,106 @@
+# Getting started
+
+To get the embedded driver for your Sensirion Sensor working there are now
+only a few steps missing. These steps are needed since our driver is made
+in a way that it is possible to run it on different hardware platforms.
+So we need to adjust two files according to your platform.
+
+
+
```
Comment (abrauchli) [inline i2c/README.md:9]:
```suggestion

```

File: i2c/README.md [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+The first one is `sensirion_i2c_hal.c`.
+In this file it is implemented how to send your requests to the Sensor.
+How this is done depends on your hardware platform therefore we can't
```
Comment (abrauchli) [inline i2c/README.md:12]:
```suggestion
How this is done depends on your hardware platform. Therefore we can't
```

File: i2c/README.md [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+provide just one version of this code. The default version of this file
+used is an empty one, where you could implement it yourself. Luckily
+for most common platforms we provide already implemented versions in
```
Comment (abrauchli) [inline i2c/README.md:15]:
```suggestion
used is an empty one, where you could implement it yourself. For
some common platforms we provide an implementation in
```

File: i2c/README.md [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+the `sample-implementations` folder.
+If you are using for example a Linux based platform like Raspberry Pi
+you can just replace the unimplemented file with the provided
```
Comment (abrauchli) [inline i2c/README.md:18]:
```suggestion
you can just replace the unimplemented HAL template with the provided
```

File: i2c/README.md [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+implementation in `sample-implementations/linux_user_space/` like so:
```
Comment (abrauchli) [inline i2c/README.md:19]:
```suggestion
implementation in `sample-implementations/linux_user_space/` by replacing the template with the actual implementation:
```

File: i2c/README.md [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+
+```
+cp sample-implementations/linux_user_space/sensirion_i2c_hal.c ./
+```
+
+If you are on a different platform just use one of the other sample
+implementations or implement `sensirion_i2c_hal.c` yourself.
+
+
+
```
Comment (abrauchli) [inline i2c/README.md:29]:
```suggestion

```

File: i2c/README.md [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+The second file is `sensirion_config.h`.
+If you are on a Linux based platform you can skip this part since
+everything is already correctly setup for you.
+Otherwise you need to check if the libraries `<stdint.h>` and
```
Comment (abrauchli) [inline i2c/README.md:33]: see changes proposed above (I2C)

File: i2c/README.md [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+* `uint16_t` = unsigned 16bit integer
+* `int8_t` = signed 8bit integer
+* `uint8_t` = unsigned 8bit integer
+
+In addition to that you will need to specify `NULL`.
+For both we have a commented template where just need to fill in
+your system specific values.
+
+
+
```
Comment (abrauchli) [inline i2c/README.md:54]:
```suggestion

```

File: i2c/README.md [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+2. Navigate the directory where this README is located.
+3. Run `make`, this compiles all the code here to one binary.
+4. Run `./[SENSORNAME]_example_usage`, where you replace `[SENSORNAME]`
+   with your sensors name. This will run your newly compiled binary.
+5. Now you should see the first measurement values appear in your terminal.
+   As a next step you can adjust the example usage file or write your own
+   main function to use the sensor.
+
+We hope this tutorial was helpful and you are now able to use the embedded drivers
+to communicate with your Sensirion Sensor.
```
Comment (abrauchli) [inline i2c/README.md:72]:
```suggestion
to communicate with your Sensirion sensor.
```

File: i2c/README.md [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+the sensor and convert is back to data readable by your machine. The functions in
+here calculate and check CRCs, reorder bytes for different byte orders and build the
+correct formatted frame for your sensor.
+
+### sensirion\_i2c\_hal.[ch]
+
+In these files you can find the implementation of the hardware abstraction layer used
+by Sensirion I2C embedded drivers. This part of the code is specific to the underlying
+hardware platform, therefore multiple implementations exist. We also provide an
+unimplemented file for the user to implement himself or herself.
```
Comment (abrauchli) [inline i2c/README.md:92]:
```suggestion
unimplemented template for the user to implement.
```

File: i2c/README.md [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+In the `sample-implementations/` folder we provide implementations for the most common
+platforms.
+
+### sensirion\_config.h
+
+In this file we keep all the included libraries for our drivers and global defines.
+Next to `sensirion_i2c_hal.c` it's the only file you should need to edit to get your
+driver working.
```
Comment (abrauchli) [inline i2c/README.md:100]:
```suggestion
Next to `sensirion_i2c_hal.c` *it's the only file you should need to edit to get your
driver working.*
```

File: i2c/sample-implementations/GPIO_bit_banging/README.md [hunk 29]
```diff
@@ -0,0 +1,15 @@
+# GPIO based Software I2C
+
+This implementation of the I2C HAL uses bit banging on GPIOs to run
+the I2C communication. It is different from other sample-implementations.
+It uses an additional pair of `.c` and `.h` file to make the implementation
+generic. These file are called `sensirion_i2c_gpio.[ch]`. Like with the I2C
```
Comment (abrauchli) [inline i2c/sample-implementations/GPIO_bit_banging/README.md:6]:
```suggestion
generic, in a way that GPIO bitbanging can be performed on any system providing access to GPIOs. Note that in this case the pull-up resistors in the electric circuit are still needed, or need to be enabled manually if your microprocessor provides this option. These file are called `sensirion_i2c_gpio.[ch]`. Like with the I2C
```

File: i2c/sample-implementations/GPIO_bit_banging/README.md [hunk 30]
```diff
@@ -0,0 +1,15 @@
...
+HAL we provide a empty file in this folder to implement yourself for your
+platform and sample implementations in the so called folder.
+
+## Getting started
+
+To use this I2C HAL copy `sensirion_i2c_hal.c` and `sensirion_i2c_gpio.h`
+from this folder to the main folder of your driver. Then either choose a
+sample implementation or implement `sensirion_i2c_gpio.c` yourself and copy
```
Comment (abrauchli) [inline i2c/sample-implementations/GPIO_bit_banging/README.md:14]:
```suggestion
sample implementation or implement `sensirion_i2c_gpio.c` and copy
```

**Review by abrauchli — CHANGES_REQUESTED** [review]

File: README.md [hunk 31]
```diff
@@ -1,7 +1,14 @@
...
-for both parts. It is a dependency of the Sensirion embedded drivers.
+This repository contains the implementation of the Sensirion I2C and SHDLC
+protocol and some commonly used functions for sensor implementations.
+It is a dependency of the Sensirion embedded sensor drivers. Sensirion
+senors all use a protocol on top of the bit level communications like I2C or
+UART. There are two of those kinds protocols used by Sensirion sensors, one
+for UART communication and one for I2C communication. These two work the same
+for all sensor types and therefore need to be implemented only once. To keep
+this implementation in one place this repository exists and is used by all
+embedded drivers.
```
Comment (Zifzaf) [inline README.md:10]:
```suggestion
senors all use a protocol on top of the I2C/UART communications layer.
There are two of those kinds protocols used by Sensirion sensors, one
for UART communication, called SHDLC, and one for I2C communication.
These are shared amongst sensor types and are therefore implemented only once. 
To keep this implementation in one place this repository exists and is used by all
embedded drivers.
```

Comment (Zifzaf) [inline README.md:10]:
```suggestion
senors all use a protocol on top of the I2C/UART communications layer.
There are two of those kinds protocols used by Sensirion sensors, one
for UART communication, called SHDLC, and one for I2C communication.
These are shared amongst sensor types and are therefore implemented only once. 
To keep this implementation in one place this repository exists and is used by all
embedded drivers.
```

File: README.md [hunk 32]
```diff
@@ -1,7 +1,14 @@
 # embedded-common
-This repository contains the Sensirion I2C hardware abstraction layer,
-the Sensirion SHDLC/UART hardware abstraction layer and commonly shared code
-for both parts. It is a dependency of the Sensirion embedded drivers.
+This repository contains the implementation of the Sensirion I2C and SHDLC
+protocol and some commonly used functions for sensor implementations.
+It is a dependency of the Sensirion embedded sensor drivers. Sensirion
+senors all use a protocol on top of the I2C/UART communications layer.
+There are two of those kinds protocols used by Sensirion sensors, one
```
Comment (abrauchli) [inline README.md:6]:
```suggestion
There are two of those kinds of protocols used by Sensirion sensors, one
```

File: i2c/README.md [hunk 33]
```diff
@@ -0,0 +1,99 @@
+# Getting started
+
+To get the embedded driver for your Sensirion Sensor working there are now
+only a few steps missing. These steps are needed since our driver is made
+in a way that it is possible to run it on different hardware platforms.
+So we need to adjust two files according to your platform.
+The first one is `sensirion_i2c_hal.c`.
+In this file it is implemented how to send your requests to the Sensor.
```
Comment (abrauchli) [inline i2c/README.md:8]:
```suggestion
In this file it is implemented how to send your requests to the sensor.
```

check if there are other instances of capitalized Sensors

**Review by abrauchli — CHANGES_REQUESTED** [review]

File: README.md [hunk 34]
```diff
@@ -35,25 +89,31 @@ specify the following types yourself:
...
-the GPIO pins used for SDA (data line) and SCL (clock line) along with the delay
-function, that must be precise enough to drive the pins at the desired
-frequency.  For further information please check out some of the sample
-implementations in the `i2c/sw/sample-implementations` folder in this
-repository.
+### sensirion\_i2c\_hal.c
+
+As already explained earlier is all the I2C related code in this repository based
+on Sensirion's I2C HAL. It is defined in `sensirion_i2c_hal.h` and implemented in
+the corresponding `.c` file, both located in the `i2c/` folder. To use the I2c part
```
Comment (rnestler) [inline README.md:102]:
```suggestion
the corresponding `.c` file, both located in the `i2c/` folder. To use the I2C part
```

File: README.md [hunk 35]
```diff
@@ -10,21 +17,68 @@ files) from the respective projects. An overview is listed at
...
+most common platforms. We even provide a implementation for software I2C using
+bit-banging on GPIOs in `i2c/sample-implementations/GPIO_bit_banging/`. If your
+platform isn't currently there, you can use the provided template structure in the
+`i2c/` folder to implement it yourself. We're very happy to review and include more
+architectures in the form of a pull request on GitHub.
+
+### SHDLC
+
+In the `shdlc` folder is the implementation of the protocol used by Sensirion
+sensors for UART communication. These sensors use Sensirions own SHDLC protocol.
```
Comment (rnestler) [inline README.md:53]:
```suggestion
sensors for UART communication. These sensors use Sensirion's own SHDLC protocol.
```

You use Sensirion's everywhere else. I'd also be fine with using Sensirions everywhere.

**Review by rnestler — COMMENTED** [review]
Small typo / consistency fixes.

File: README.md [hunk 36]
```diff
@@ -35,25 +89,31 @@ specify the following types yourself:
...
-implementations in the `i2c/sw/sample-implementations` folder in this
-repository.
+### sensirion\_i2c\_hal.c
+
+As already explained earlier is all the I2C related code in this repository based
+on Sensirion's I2C HAL. It is defined in `sensirion_i2c_hal.h` and implemented in
+the corresponding `.c` file, both located in the `i2c/` folder. To use the I2c part
+of `embedded-common` you either need to implement the `sensirion_i2c_hal.c` file or
+replace the unimplemented version in `i2c/` with the correct sample implementation
+from `i2c/sample-implementation/`.
```
Comment (rnestler) [inline README.md:105]: I found this sentence a bit confusing: Maybe better: "you either need to implement the functions in `sensirion_i2c_hal.c` or replace it with a sample implementation from `i2c/sample-implementation/`".

Comment (rnestler) [inline README.md:105]:
```suggestion
of `embedded-common` you either need to implement the functions in sensirion_i2c_hal.c or
replace it with a sample implementation from i2c/sample-implementation/.
```

File: README.md [hunk 37]
```diff
@@ -2,11 +2,11 @@
...
 It is a dependency of the Sensirion embedded sensor drivers. Sensirion
-senors all use a protocol on top of the bit level communications like I2C or
-UART. There are two of those kinds protocols used by Sensirion sensors, one
-for UART communication and one for I2C communication. These two work the same
-for all sensor types and therefore need to be implemented only once. To keep
-this implementation in one place this repository exists and is used by all
+sensors all use a protocol on top of the I2C/UART communications layer.
+There are two of those kinds of protocols used by Sensirion sensors, one
+for UART communication, called SHDLC, and one for I2C communication.
+These are shared amongst sensor types and are therefore implemented only once.
```
Comment (chjaeggi) [inline README.md:8]: no need for second are... suggestion leave it away completely

File: README.md [hunk 38]
```diff
@@ -2,11 +2,11 @@
...
+To keep this implementation in one place this repository exists and is used by all
```
Comment (chjaeggi) [inline README.md:9]: ... in one place only, this repository ...

File: README.md [hunk 39]
```diff
@@ -18,48 +18,51 @@ https://github.com/Sensirion/info#repositories
...
 ### Common
 
-In the `common/` folder are three files. `senirion_config.h` is a general
+In the `common/` folder are three files. `senirion_config.h` is the general
 configuration file. Depending on your hardware you will need to edit this.
-You can find more about this further down. There are also
-`sensirion_common.[ch]`. These contain common helper functions used by both
-UART and I2C sensor drivers.
+For more information, see the section named `sensirion_config.h` further down
+in this document. There are also `sensirion_common.[ch]`. These contain common
```
Comment (chjaeggi) [inline README.md:29]: There are also the two files ...

File: README.md [hunk 40]
```diff
@@ -18,48 +18,51 @@ https://github.com/Sensirion/info#repositories
...
+swapping, ...) of Sensirion's I2C communication are implemented. The finished frame
+is then sent to the attached hardware using the Hardware Abstraction Layer (HAL) in
+the other two files. Since the way how to send out this frame depends on underlying
+hardware and operating system, we provide different implementations for that part
 of the code. You can find these in the folder `i2c/sample-implementations/` for the
 most common platforms. We even provide a implementation for software I2C using
 bit-banging on GPIOs in `i2c/sample-implementations/GPIO_bit_banging/`. If your
-platform isn't there you can use the empty structure in the `i2c/` folder to
-implement it yourself.
+platform isn't currently there, you can use the provided template structure in the
```
Comment (chjaeggi) [inline README.md:46]: If we don't provide code for your platform yet you can use ...

File: README.md [hunk 41]
```diff
@@ -18,48 +18,51 @@ https://github.com/Sensirion/info#repositories
...
+`i2c/` folder to implement it yourself. We're very happy to review and include more
+architectures in the form of a pull request on GitHub.
 
 ### SHDLC
 
 In the `shdlc` folder is the implementation of the protocol used by Sensirion
```
Comment (chjaeggi) [inline README.md:52]: The shdlc folder contains ...

File: README.md [hunk 42]
```diff
@@ -18,48 +18,51 @@ https://github.com/Sensirion/info#repositories
...
-(Hardware Abstraction Layer) in the other two files. Since the way how to send
+byte order swapping, frame construction, ...) of Sensirion's SHDLC communication are
+implemented. The finished frame is then sent to your hardware using the HAL
+in the other two files. Since the way how to send
 out this frame depends on underlying hardware and operating system we provide
 different implementations for that part of the code. You can find these in the
 folder `shdlc/sample-implementaions/` for the most common platforms. If your
-platform isn't there you can use the empty structure in the `shdlc/` folder to
-implement it yourself.
+platform isn't currently there, you can use the provided template structure in the
```
Comment (chjaeggi) [inline README.md:63]: see above

File: i2c/README.md [hunk 43]
```diff
@@ -4,39 +4,35 @@ To get the embedded driver for your Sensirion Sensor working there are now
 only a few steps missing. These steps are needed since our driver is made
 in a way that it is possible to run it on different hardware platforms.
 So we need to adjust two files according to your platform.
-
-
-
 The first one is `sensirion_i2c_hal.c`.
-In this file it is implemented how to send your requests to the Sensor.
-How this is done depends on your hardware platform therefore we can't
+In this file it is implemented how to send your requests to the sensor.
```
Comment (chjaeggi) [inline i2c/README.md:8]: This file contains the implementation of the sensor communication (how to send requests to it)

File: i2c/README.md [hunk 44]
```diff
@@ -4,39 +4,35 @@ To get the embedded driver for your Sensirion Sensor working there are now
...
+How this is done depends on your hardware platform. Therefore we can't
```
Comment (chjaeggi) [inline i2c/README.md:9]: Therefore, ...

File: i2c/README.md [hunk 45]
```diff
@@ -4,39 +4,35 @@ To get the embedded driver for your Sensirion Sensor working there are now
...
 provide just one version of this code. The default version of this file
-used is an empty one, where you could implement it yourself. Luckily
-for most common platforms we provide already implemented versions in
+used is an empty one, where you could implement it yourself. For
```
Comment (chjaeggi) [inline i2c/README.md:11]: I don't understand this sentence: `The default version of this file used is an empty one, where you could implement it yourself.` So it's an empty file we provide? That means the user needs to implement it not `could`, right?

File: i2c/README.md [hunk 46]
```diff
@@ -89,15 +82,15 @@ correct formatted frame for your sensor.
 In these files you can find the implementation of the hardware abstraction layer used
 by Sensirion I2C embedded drivers. This part of the code is specific to the underlying
 hardware platform, therefore multiple implementations exist. We also provide an
-unimplemented file for the user to implement himself or herself.
+unimplemented template for the user to implement.
```
Comment (chjaeggi) [inline i2c/README.md:85]: This sounds much clearer to me compared to my previous comment. Can we write a similar instruction there?

File: i2c/sample-implementations/GPIO_bit_banging/README.md [hunk 47]
```diff
@@ -3,13 +3,16 @@
 This implementation of the I2C HAL uses bit banging on GPIOs to run
 the I2C communication. It is different from other sample-implementations.
 It uses an additional pair of `.c` and `.h` file to make the implementation
-generic. These file are called `sensirion_i2c_gpio.[ch]`. Like with the I2C
-HAL we provide a empty file in this folder to implement yourself for your
-platform and sample implementations in the so called folder.
+generic, in a way that GPIO bitbanging can be performed on any system providing
+access to GPIOs. Note that in this case the pull-up resistors in the electric
+circuit are still needed, or need to be enabled manually if your microprocessor
+provides this option. These file are called `sensirion_i2c_gpio.[ch]`.
```
Comment (chjaeggi) [inline i2c/sample-implementations/GPIO_bit_banging/README.md:9]: files

File: i2c/sample-implementations/GPIO_bit_banging/README.md [hunk 48]
```diff
@@ -3,13 +3,16 @@
...
+Like with the I2C HAL we provide a empty file in this folder to implement
```
Comment (chjaeggi) [inline i2c/sample-implementations/GPIO_bit_banging/README.md:10]: Similar to the I2C HAL, we provide an empty template(?) in this ...

File: shdlc/README.md [hunk 49]
```diff
@@ -4,39 +4,36 @@ To get the embedded driver for your Sensirion Sensor working there are now
 only a few steps missing. These steps are needed since our driver is made
 in a way that it is possible to run it on different hardware platforms.
 So we need to adjust two files according to your platform.
-
-
-
 The first one is `sensirion_uart_hal.c`.
 In this file it is implemented how to send your requests to the Sensor.
-How this is done depends on your hardware platform therefore we can't
+How this is done depends on your hardware platform. Therefore we can't
```
Comment (chjaeggi) [inline shdlc/README.md:9]: Therefore, ...

File: shdlc/README.md [hunk 50]
```diff
@@ -4,39 +4,36 @@ To get the embedded driver for your Sensirion Sensor working there are now
...
 provide just one version of this code. The default version of this file
-used is an empty one, where you could implement it yourself. Luckily
-for most common platforms we provide already implemented versions in
-the `sample-implementations` folder.
+used is an empty one, where you could implement it yourself. For some
```
Comment (chjaeggi) [inline shdlc/README.md:11]: mentioned earlier my confusion

**Review by chjaeggi — CHANGES_REQUESTED** [review]

File: i2c/README.md (hunk 45 above)
Comment (Zifzaf) [reply i2c/README.md:11]: Is this better? The default version of this file is an empty template, where you can implement it yourself.

**Review by rnestler — APPROVED** [review]

**Pull request merged**